text/utilities/0000-health-checker-buildpack.md

```diff
@@ -0,0 +1,70 @@
+
+Having a buildpack that installs a health checker allows for HTTP-based health checks when running on Docker. In addition, it could allow for other types of health checkers for example if your application does not use HTTP.
+
+## Detailed Explanation
+
+1. A health checker buildpack has been developed. It will be contributed to Paketo.
+   1. By default, no health checker will be installed by the buildpack.
+   2. If a user sets `BP_HEALTH_CHECKER_ENABLED=true` then the buildpack will detect and contribute a health check binary to the `$PATH`.
+   3. A process type of `health-check` will be contributed with the command to run the health checker. This may be invoked by the user rather than needing the full command.
+
+## Rationale and Alternatives
+
+1. A user can create a custom stack and include a tool like `curl` or `wget`. This is a lot of additional work, plus `curl` and `wget` are overkill for a simple HTTP health check. To reduce footprint, they are not recommended.
+
+## Implementation
+
+The new Health Checker buildpack will participate if all the following conditions are met
+
+* A user sets `BP_HEALTH_CHECKER_ENABLED=true`
+
+When conditions are met, the buildpack will do the following:
+
+* Contributes a health checker to a layer marked `launch` with command on `$PATH`
+* Contributes a process type called `health-check` which will execute the installed health checker.
+
+The Health Checker buildpack will optionally be added to the language family buildpacks that wish to include support for it. The primary targets will be the Java and Rust buildpacks.
+
+### Example
+
+A user can then utilize the health checker like this:
+
+1. For a Spring Boot app, run `pack build apps/maven -p target/demo-0.0.1-SNAPSHOT.jar -b paketo-buildpacks/java -b docker.io/paketo-buildpacks/health-checker -e BP_HEALTH_CHECKER_ENABLED=true -e BP_HEALTH_CHECKER_THC_ARGS='8080 /actuator/health'` (it won't be necessary to include the `-b` arguments once the buildpack has been integrated with language family buildpacks). You'll see output like this at the end. It's installing the health checker and setting a process type.
+
+    ```
+    Paketo Health Checker Buildpack vDEVELOPMENT
+      https://github.com/paketo-buildpacks/health-checker
+      Build Configuration:
+        $BP_HEALTH_CHECKER_DEPENDENCY  thc                    which health checker to contribute
+        $BP_HEALTH_CHECKER_ENABLED     true                   contributes a health checker if enabled
+        $BP_HEALTH_CHECKER_THC_ARGS    8080 /actuator/health  arguments passed to tiny-health-checker, if used
```

Comment:
That's a fair criticism. I think the rationale for positional args here is this:

1. I wanted the smallest static binary as possible, so not pulling in a library to parse arguments helps keep the size as low as possible. Right now, it's super simple to pull the positional arguments.

2. If we take the route of having support for multiple health checkers in the buildpack, that keeps things flexible as well. We can have `thc` essentially be feature complete. It is just doing HTTP health checks, it does that, so it's done (maybe the method should be configurable? support for HEAD?). We can then add another different checker for TCP, a different one for gRPC, a different one for RSocket, or whatever else we need to support.

If all else fails, we can expose additional functionality through env variable settings, rather than continuing to add arguments. It might even make sense to do this with PORT & PATH, not that we strictly need it at the moment.

Reply:
> I wanted the smallest static binary as possible, so not pulling in a library to parse arguments helps keep the size as low as possible. Right now, it's super simple to pull the positional arguments.

We could use the [flag package](https://pkg.go.dev/flag) from the go standard library which is pretty lean, if not as flexible/pretty as something like cobra